sphinx/util/inspect.py

```diff
@@ -315,6 +317,76 @@ def is_builtin_class_method(obj: Any, attr_name: str) -> bool:
     return getattr(builtins, safe_getattr(cls, '__name__', '')) is cls
 
 
+def iter_args(func: Union[ast.FunctionDef, ast.AsyncFunctionDef]
+              ) -> Generator[str, None, None]:
+    """Get an iterator for arguments names from FunctionDef node."""
+    if hasattr(func.args, "posonlyargs"):  # py38 or above
+        yield from (a.arg for a in func.args.posonlyargs)  # type: ignore
+    yield from (a.arg for a in func.args.args)
+    if func.args.vararg:
+        yield func.args.vararg.arg
+    if func.args.kwarg:
+        yield func.args.kwarg.arg
+
+
+def parse_argtypes(s: str) -> Generator[str, None, None]:
+    """Parse argument part of type_comment."""
+    start = 0
+    parens = 0
+    for i, char in enumerate(s):
+        if char == '[':
+            parens += 1
+        elif char == ']':
+            parens -= 1
+        elif char == ',' and parens == 0:
+            yield s[start:i]
+            start = i + 1
+
+    yield s[start:]
+
+
+def get_type_hints_from_type_comment(obj: Any) -> Dict[str, str]:
+    """Get type hints from py2 style type_comment.
+
+    Python3.8+ or typed_ast is required.
+    """
+    if sys.version_info > (3, 8):
+        parse = partial(ast.parse, type_comments=True)
+    else:
+        try:
+            from typed_ast import ast3
+            parse = ast3.parse
+        except ImportError:
+            return {}
+
+    try:
+        source = inspect.getsource(obj)
+        if source.startswith((' ', r'\t')):
+            # subject is placed inside class or block.  To read its docstring,
+            # this adds if-block before the declaration.
+            module = parse('if True:\n' + source)
+            subject = module.body[0].body[0]  # type: ignore
+        else:
+            module = parse(source)
+            subject = module.body[0]  # type: ignore
+        type_comment = subject.type_comment
+
+        if type_comment is None:  # no type_comment
+            return {}
+        else:
+            type_hints = {}  # type: Dict[str, Any]
+            argtypes, rtype = type_comment_re.match(type_comment).groups()
```

Comment:
Instead of using string parsing, can you use `ast.parse(type_comment, mode='func_type')`?

Reply:
Thank you for the info. I'll take a look it tomorrow!

Reply:
Fully rewritten with `mode=func_type` now!